### PR TITLE
ITP Protocol and Basic Replication Integration

### PR DESCRIPTION
**Pull request by tpan496**
Installs basic replication code from Gus to the terrier master. 
- Basic ITP protocol that involves `StartReplication` and `StopReplication` commands.
- Changes to the log manager and networking layer to incorporate replication logs.
- A basic `CreateDatabase` test where the database oid of the master and replica should match.

### CODE REVIEW

**Comment by tpan496** [comment]
@lmwnshn finished code cleanup, ready for review. thanks!

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -12,7 +12,7 @@ list(REMOVE_ITEM TERRIER_SRCS ${PROJECT_SOURCE_DIR}/src/main/terrier.cpp)
 # spdlog
 file(GLOB_RECURSE SPDLOG_SOURCES
         ${PROJECT_SOURCE_DIR}/third_party/spdlog/*.h
-)
+        )
```
Comment (lmwnshn) [inline src/CMakeLists.txt:15]: Nit: unnecessary change.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -22,7 +22,7 @@ add_subdirectory(${PROJECT_SOURCE_DIR}/third_party/libpg_query/ libpg_query.a)
 file(GLOB_RECURSE BWTREE_SOURCES
         ${PROJECT_SOURCE_DIR}/third_party/bwtree/*.cpp
         ${PROJECT_SOURCE_DIR}/third_party/bwtree/*.h
-)
+        )
```
Comment (lmwnshn) [inline src/CMakeLists.txt:25]: Nit: unnecessary change.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -32,14 +32,14 @@ list(APPEND TERRIER_SRCS ${PROJECT_SOURCE_DIR}/third_party/nlohmann/json.hpp)
 file(GLOB_RECURSE MADOKA_SOURCES
         ${PROJECT_SOURCE_DIR}/third_party/madoka/*.cc
         ${PROJECT_SOURCE_DIR}/third_party/madoka/*.h
-)
+        )
```
Comment (lmwnshn) [inline src/CMakeLists.txt:35]: Nit: unnecessary change.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -32,14 +32,14 @@ list(APPEND TERRIER_SRCS ${PROJECT_SOURCE_DIR}/third_party/nlohmann/json.hpp)
...
 list(APPEND TERRIER_SRCS ${MADOKA_SOURCES})
 
 # libcount
 file(GLOB_RECURSE LIBCOUNT_SOURCES
         ${PROJECT_SOURCE_DIR}/third_party/libcount/*.cc
         ${PROJECT_SOURCE_DIR}/third_party/libcount/*.h
-)
+        )
```
Comment (lmwnshn) [inline src/CMakeLists.txt:42]: Nit: unnecessary change.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -70,11 +70,11 @@ ADD_TERRIER_LIB(terrier
         STATIC_LINK_LIBS ${TERRIER_STATIC_LINK_LIBS}
         STATIC_PRIVATE_LINK_LIBS ${TERRIER_STATIC_PRIVATE_LINK_LIBS}
         DEPENDENCIES ${TERRIER_DEPENDENCIES}
-)
+        )
```
Comment (lmwnshn) [inline src/CMakeLists.txt:73]: Nit: unnecessary change.

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -70,11 +70,11 @@ ADD_TERRIER_LIB(terrier
...
 
 ###############################################
 # Terrier executable
 ###############################################
-add_executable(terrier main/terrier.cpp)
+add_executable(terrier main/terrier.cpp include/storage/write_ahead_log/replication_log_consumer_task.h storage/write_ahead_log/replication_log_consumer_task.cpp include/storage/write_ahead_log/log_consumer_task.h)
```
Comment (lmwnshn) [inline src/CMakeLists.txt:78]: Nit: this should not be necessary (and might actually be incorrect).

File: src/include/main/db_main.h [hunk 7]
```diff
@@ -240,6 +241,8 @@ class DBMain {
       server_ = std::make_unique<network::TerrierServer>(common::ManagedPointer(provider_),
                                                          common::ManagedPointer(connection_handle_factory_),
                                                          thread_registry, port, connection_thread_count);
+      server_->RegisterProtocol(port, common::ManagedPointer(provider_), CONNECTION_THREAD_COUNT,
```
Comment (lmwnshn) [inline src/include/main/db_main.h:244]: Does this logic belong in the network layer? Or does it belong in TerrierServer? IMO it belongs in TerrierServer. Also, if you move it to TerrierServer you don't need to include settings.h.

File: src/include/network/itp/itp_packet_writer.h [hunk 8]
```diff
@@ -27,22 +27,24 @@ class ITPPacketWriter : public PacketWriter {
...
+    BeginPacket(NetworkMessageType::ITP_REPLICATION_COMMAND).AppendValue(message_id).AppendValue(data_size);
+  }
 
   /**
    * End the Replication command
    */
   void EndReplicationCommand() { EndPacket(); }
 
   /**
-   * Writes a Stop Replication packet
```
Comment (lmwnshn) [inline src/include/network/itp/itp_packet_writer.h:38]: Nit: it seems like this change does not do anything except reorder the code.

File: src/include/network/itp/itp_protocol_interpreter.h [hunk 9]
```diff
@@ -85,7 +85,12 @@ class ITPProtocolInterpreter : public ProtocolInterpreter {
   void SetPacketMessageType(common::ManagedPointer<ReadBuffer> in) override;
 
  private:
+  InputPacket curr_input_packet_{};
```
Comment (lmwnshn) [inline src/include/network/itp/itp_protocol_interpreter.h:88]: Please add a comment describing this.

File: src/include/network/itp/itp_protocol_interpreter.h [hunk 10]
```diff
@@ -85,7 +85,12 @@ class ITPProtocolInterpreter : public ProtocolInterpreter {
...
+  std::unordered_map<std::string, std::string> cmdline_options_;
```
Comment (lmwnshn) [inline src/include/network/itp/itp_protocol_interpreter.h:89]: Please add a comment describing this + I haven't seen how this is used yet, but I'm suspicious of keeping cmdline options here. It should probably be a setting.

File: src/include/network/network_defs.h [hunk 11]
```diff
@@ -16,6 +16,9 @@ class ReadBuffer;
 // want anyone using it to directly access the socket downstream
 STRONG_TYPEDEF_HEADER(connection_id_t, uint16_t);
 
+// For threads
+#define CONNECTION_THREAD_COUNT 4
```
Comment (lmwnshn) [inline src/include/network/network_defs.h:20]: It is not clear to me what this is doing, can you elaborate on the comment?

Comment (lmwnshn) [reply src/include/network/network_defs.h:20]: We already have connection_thread_count being passed in to TerrierServer, why does RegisterProtocol need its own count? Are they related?

File: src/include/network/network_io_utils.h [hunk 12]
```diff
@@ -221,6 +221,13 @@ class ReadBufferView {
     return result;
   }
 
+  /**
+   * @param bytes The amount of bytes between the current cursor and rest of the buffer view
+   * @return Whether there is any more bytes between the cursor and
+   *         the end of the buffer view
+   */
+  bool HasMore(size_t bytes = 1) { return offset_ + bytes <= size_; }
```
Comment (lmwnshn) [inline src/include/network/network_io_utils.h:229]: This is identical to Buffer::HasMore, but I understand that you might need similar functionality in the ReadBufferView.

That said, we're trying to move away from having default arguments. Consider rewriting to `HasMore(size_t bytes)` and maybe if you really hate calling `HasMore(1)` then you can have an `IsFinished` or something like that which calls `HasMore(1)` for you.

File: src/include/network/network_io_wrapper.h [hunk 13]
```diff
@@ -37,6 +39,25 @@ class NetworkIoWrapper {
     RestartState();
   }
 
+  explicit NetworkIoWrapper(const std::string &ip_address, uint16_t port)
+      : sock_fd_(socket(AF_INET, SOCK_STREAM, 0)),
```
Comment (lmwnshn) [inline src/include/network/network_io_wrapper.h:43]: This does not check the error code of socket.

Comment (lmwnshn) [reply src/include/network/network_io_wrapper.h:43]: In general, this function seems to be a hack. Is there a reason to not create the socket from whatever is calling this?

File: src/include/network/network_io_wrapper.h [hunk 14]
```diff
@@ -37,6 +39,25 @@ class NetworkIoWrapper {
...
+        in_(std::make_unique<ReadBuffer>()),
+        out_(std::make_unique<WriteQueue>()) {
+    struct sockaddr_in serv_addr;
+    memset(&serv_addr, 0, sizeof(serv_addr));
+    serv_addr.sin_family = AF_INET;
+    serv_addr.sin_addr.s_addr = inet_addr(ip_address.c_str());
+    serv_addr.sin_port = htons(port);
+
+    int64_t ret UNUSED_ATTRIBUTE = connect(sock_fd_, reinterpret_cast<sockaddr *>(&serv_addr), sizeof(serv_addr));
+    // TODO(Gus): we need better exception handling here
```
Comment (lmwnshn) [inline src/include/network/network_io_wrapper.h:53]: Gus is graduated, assign TODOs to me (WAN) or yourself (whatever you use).

File: src/include/network/protocol_interpreter.h [hunk 15]
```diff
@@ -90,7 +90,7 @@ class ProtocolInterpreter {
    * @param in The ReadBuffer to read input from
    * @return whether the packet header is valid or not
    */
-  bool TryReadPacketHeader(const common::ManagedPointer<ReadBuffer> in) {
+  virtual bool TryReadPacketHeader(const common::ManagedPointer<ReadBuffer> in) {
```
Comment (lmwnshn) [inline src/include/network/protocol_interpreter.h:93]: Is there a reason to make this virtual?

File: src/network/itp/itp_protocol_interpreter.cpp [hunk 16]
```diff
@@ -40,4 +42,54 @@ void ITPProtocolInterpreter::SetPacketMessageType(const common::ManagedPointer<R
...
+  curr_input_packet_.msg_type_ = in->ReadValue<NetworkMessageType>();
+  curr_input_packet_.len_ = in->ReadValue<uint32_t>() - sizeof(uint32_t);
+  if (curr_input_packet_.len_ > PACKET_LEN_LIMIT) {
+    NETWORK_LOG_ERROR("Packet size {} > limit {}", curr_input_packet_.len_, PACKET_LEN_LIMIT);
+    throw NETWORK_PROCESS_EXCEPTION("Packet too large");
+  }
+
+  // Extend the buffer as needed
+  if (curr_input_packet_.len_ > in->Capacity()) {
+    // Allocate a larger buffer to copy bytes off from the I/O layer's buffer
```
Comment (lmwnshn) [inline src/network/itp/itp_protocol_interpreter.cpp:82]: As far as I can tell, this is the only change from the original function.

File: src/network/itp/itp_protocol_interpreter.cpp [hunk 17]
```diff
@@ -40,4 +42,54 @@ void ITPProtocolInterpreter::SetPacketMessageType(const common::ManagedPointer<R
   curr_input_packet_.msg_type_ = in->ReadValue<NetworkMessageType>();
 }
 
+bool ITPProtocolInterpreter::TryBuildPacket(common::ManagedPointer<ReadBuffer> in) {
+  if (!TryReadPacketHeader(in)) return false;
```
Comment (lmwnshn) [inline src/network/itp/itp_protocol_interpreter.cpp:46]: As far as I can tell, this is identical to the existing function.

File: src/include/network/protocol_interpreter.h (hunk 15 above)
Comment (lmwnshn) [reply src/include/network/protocol_interpreter.h:93]: After looking at the new implementations, I do not think this should be made virtual. I think `TryBuildPacket`'s override is identical to the original, and `TryReadPacketHeader`'s override is almost identical except for a copy of the buffer (which could happen elsewhere).

File: src/include/storage/recovery/replication_log_provider.h [hunk 18]
```diff
@@ -1,19 +1,110 @@
 #pragma once
 
+#include <queue>
 #include <memory>
+#include <utility>
+
+#include <string>
```
Comment (lmwnshn) [inline src/include/storage/recovery/replication_log_provider.h:7]: Surprised format isn't grouping these together, or maybe you need to run that.

File: src/include/storage/write_ahead_log/log_manager.h [hunk 19]
```diff
@@ -57,15 +58,18 @@ class LogManager : public common::DedicatedThreadOwner {
   LogManager(std::string log_file_path, uint64_t num_buffers, std::chrono::microseconds serialization_interval,
              std::chrono::microseconds persist_interval, uint64_t persist_threshold,
              common::ManagedPointer<RecordBufferSegmentPool> buffer_pool,
-             common::ManagedPointer<terrier::common::DedicatedThreadRegistry> thread_registry)
+             common::ManagedPointer<terrier::common::DedicatedThreadRegistry> thread_registry,
+             const std::string& ip_address = "", uint16_t replication_port = 0)
```
Comment (lmwnshn) [inline src/include/storage/write_ahead_log/log_manager.h:62]: Avoid default args, especially if they can be buggy.

File: src/network/terrier_server.cpp [hunk 20]
```diff
@@ -42,41 +42,43 @@ void TerrierServer::RunServer() {
...
-  int retval = bind(listen_fd_, reinterpret_cast<struct sockaddr *>(&sin), sizeof(sin));
-  if (retval < 0) {
-    NETWORK_LOG_ERROR("Failed to bind socket: {}", strerror(errno));
-    throw NETWORK_PROCESS_EXCEPTION("Failed to bind socket.");
-  }
-  retval = listen(listen_fd_, conn_backlog);
-  if (retval < 0) {
-    NETWORK_LOG_ERROR("Failed to create listen socket: {}", strerror(errno));
-    throw NETWORK_PROCESS_EXCEPTION("Failed to create listen socket.");
+  // int conn_backlog = common::Settings::CONNECTION_BACKLOG;
```
Comment (lmwnshn) [inline src/network/terrier_server.cpp:45]: Dead code.

File: test/storage/replication_test.cpp [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+  storage::ReplicationLogProvider *replica_log_provider_;
+  storage::RecoveryManager *replica_recovery_manager_;
+  network::ITPCommandFactory *replica_itp_command_factory_;
+  network::ITPProtocolInterpreter::Provider *replica_itp_protocol_provider_;
+  network::ConnectionHandleFactory *replica_connection_handle_factory_;
+  trafficcop::TrafficCop *replica_tcop_;
+  network::TerrierServer *replica_server_;
+
+  void SetUp() override {
+    TerrierTest::SetUp();
```
Comment (lmwnshn) [inline test/storage/replication_test.cpp:97]: This is an "old style" test prior to DBMain existing, take a look at how tests are brought up now. We don't need to manually create all the components, I think.

File: test/CMakeLists.txt [hunk 22]
```diff
@@ -4,7 +4,7 @@ file(GLOB_RECURSE TEST_UTIL_SRCS ${PROJECT_SOURCE_DIR}/test/test_util/*.cpp ${PR
 # Test library
 ###############################################
 if (TERRIER_BUILD_TESTS OR TERRIER_BUILD_BENCHMARKS)
-    add_library(test_util STATIC ${TEST_UTIL_SRCS})
+    add_library(test_util STATIC ${TEST_UTIL_SRCS} storage/replication_test.cpp)
```
Comment (lmwnshn) [inline test/CMakeLists.txt:7]: Nit: should not be necessary.

File: src/include/storage/write_ahead_log/replication_log_consumer_task.h [hunk 23]
```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <utility>
+#include <vector>
+#include <string>
+#include <memory>
+
+#include "network/network_io_utils.h"
```
Comment (lmwnshn) [inline src/include/storage/write_ahead_log/replication_log_consumer_task.h:8]: Nit: Some of these includes can probably be forward declarations.

File: src/include/storage/write_ahead_log/log_consumer_task.h [hunk 24]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <condition_variable>  // NOLINT
+#include <utility>
+#include <vector>
+
+#include "common/container/concurrent_blocking_queue.h"
```
Comment (lmwnshn) [inline src/include/storage/write_ahead_log/log_consumer_task.h:7]: Nit: most of these includes should be forward declarations.

File: src/network/itp/itp_protocol_interpreter.cpp [hunk 25]
```diff
@@ -20,7 +20,9 @@ Transition ITPProtocolInterpreter::Process(common::ManagedPointer<ReadBuffer> in
     NETWORK_LOG_ERROR("Encountered exception {0} when parsing packet", e.what());
     return Transition::TERMINATE;
   }
-  auto command = command_factory_->PacketToCommand(common::ManagedPointer(&curr_input_packet_));
+
+  std::shared_ptr<ITPNetworkCommand> command =
```
Comment (lmwnshn) [inline src/network/itp/itp_protocol_interpreter.cpp:24]: Nit: shared pointers tend to confuse ownership. Who owns this command?

File: src/network/terrier_server.cpp [hunk 26]
```diff
@@ -42,41 +42,43 @@ void TerrierServer::RunServer() {
...
+
+  for (auto &protocol : protocols_) {
```
Comment (lmwnshn) [inline src/network/terrier_server.cpp:47]: Nit: this logic probably should be broken out into a function. Additionally, this seems to duplicate existing code.

**Review by lmwnshn — CHANGES_REQUESTED** [review]
Per our meeting 2020-09-10, we probably will not merge this code, so you don't need to address the comments. I did a quick review though so that you can see what some issues were.

This gives us have a sense of what we had. Later, when we meet with Ricky we should get a sense of where we want to go.
